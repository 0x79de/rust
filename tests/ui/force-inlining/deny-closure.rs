//@ build-fail
//@ compile-flags: --crate-type=lib
#![allow(internal_features)]
#![feature(rustc_attrs)]

// Test that forced inlining into closures w/ errors works as expected.

#[rustc_no_mir_inline]
#[rustc_force_inline]
pub fn callee() {
}

#[rustc_no_mir_inline]
#[rustc_force_inline = "the test requires it"]
pub fn callee_justified() {
}

pub fn caller() {
    (|| {
        callee();
//~^ ERROR `callee` could not be inlined into `caller::{closure#0}` but is required to be inlined

        callee_justified();
//~^ ERROR `callee_justified` could not be inlined into `caller::{closure#0}` but is required to be inlined
    })();
}
