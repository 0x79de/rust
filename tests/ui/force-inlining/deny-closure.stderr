error: `callee` could not be inlined into `caller::{closure#0}` but is required to be inlined
  --> $DIR/deny-closure.rs:20:9
   |
LL |         callee();
   |         ^^^^^^^^ ...`callee` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]

error: `callee_justified` could not be inlined into `caller::{closure#0}` but is required to be inlined
  --> $DIR/deny-closure.rs:23:9
   |
LL |         callee_justified();
   |         ^^^^^^^^^^^^^^^^^^ ...`callee_justified` called here
   |
   = note: could not be inlined due to: #[rustc_no_mir_inline]
   = note: `callee_justified` is required to be inlined to: the test requires it

note: the above error was encountered while instantiating `fn caller::{closure#0}`
  --> $DIR/deny-closure.rs:19:5
   |
LL | /     (|| {
LL | |         callee();
...  |
LL | |     })();
   | |________^

error: aborting due to 2 previous errors

